base-R.qmd

```diff
@@ -0,0 +1,502 @@
+
+    ```{r}
+    x <- c("one", "two", "three", "four", "five")
+    x[c(3, 2, 5)]
+    ```
+
+    By repeating a position, you can actually make a longer output than input, making the term subsetting a bit of a misnomer.
+
+    ```{r}
+    x[c(1, 1, 5, 5, 5, 2)]
+    ```
+
+2.  **A vector of negative integers**.
+    Negative values drop the elements at the specified positions:
+
+    ```{r}
+    x[c(-1, -3, -5)]
+    ```
+
+3.  **A logical vector**.
+    Subsetting with a logical vector keeps all values corresponding to a `TRUE` value.
+    This is most often useful in conjunction with the comparison functions.
+
+    ```{r}
+    x <- c(10, 3, NA, 5, 8, 1, NA)
+
+    # All non-missing values of x
+    x[!is.na(x)]
+
+    # All even (or missing!) values of x
+    x[x %% 2 == 0]
+    ```
+
+    Note that, unlike `filter()`, `NA` indices will be included in the output as `NA`s.
+
+4.  **A character vector**.
+    If you have a named vector, you can subset it with a character vector:
+
+    ```{r}
+    x <- c(abc = 1, def = 2, xyz = 5)
+    x[c("xyz", "def")]
+    ```
+
+    As with positive integers, you can use a character vector to duplicate individual entries.
+
+5.  **Nothing**.
+    The final type of subsetting is nothing, `x[]`, which returns the complete `x`.
+    This is not useful for subsetting vectors, but as well see shortly it is useful when subsetting 2d structures like tibbles.
+
+### Subsetting data frames
+
+There are quite a few different ways[^base-r-1] that you can use `[` with a data frame, but the most important way is to selecting rows and columns independently with `df[rows, cols]`. Here `rows` and `cols` are vectors as described above.
+For example, `df[rows, ]` and `df[, cols]` select just rows or just columns, using the empty subset to preserve the other dimension.
+
+[^base-r-1]: Read <https://adv-r.hadley.nz/subsetting.html#subset-multiple> to see how you can also subset a data frame like it's a 1d object and how you can subset it with a matrix.
+
+Here are a couple of examples:
+
+```{r}
+df <- tibble(x = 1:3, y = c("a", "e", "f"), z = runif(3))
+
+# Select first row and second column
+df[1, 2]
+
+# Select all rows and columns x and y
+df[, c("x" , "y")]
+
+# Select rows where `x` is greater than 1 and all columns
+df[df$x > 1, ]
+```
+
+We'll come back to `$` shortly, but we need to use it here because `[` doesn't use tidy evaluation, so you need to be explicit about the source of the `x` variable.
+
+There's an important difference between tibbles and data frames when it comes to `[`.
+If `df` is a data frame, `df[, cols]` will return a vector if `col` selects a single column and a data frame if it selects more than one column.
+If `df` is a tibble, then `[` will always return a tibble.
+
+```{r}
+df1 <- data.frame(x = 1:3)
+df1[, "x"]
+
+df2 <- tibble(x = 1:3)
+df2[, "x"]
+```
+
+You can also avoid this problem by using 1d `[` to select columns:
+
+```{r}
+df1["x"]
+```
```

Comment:
I guess this is easier to explain than `drop = FALSE`?

Reply:
May be worth a sentence explicitly acknowledging the difference between `x[cols]` and `x[rows,]`